lsp/nls/src/position.rs

```diff
@@ -0,0 +1,162 @@
+}
+
+impl PositionLookup {
+    /// Create a position lookup table for looking up subterms of `rt` based on their positions.
+    pub fn new(rt: &RichTerm) -> Self {
+        let mut all_ranges = Vec::new();
+
+        rt.traverse_ref(&mut |term: &RichTerm| {
+            if let TermPos::Original(pos) = &term.pos {
+                all_ranges.push((pos.start.0, pos.end.0, term.clone()));
+            }
+            TraverseControl::<()>::Continue
+        });
+
+        // We rely on the invariant that if two ranges overlap then one is contained in the
+        // other. That is, the ranges form a tree. This sort order corresponds to pre-order
+        // traversal of that tree.
+        //
+        // (It would be nice if we could build the lookup table directly from the traversal
+        // in traverse_ref, but (1) the state tracking is fiddly and (2) the traversal order
+        // in traverse_ref can be weird; for example, a `Term::Annotated` sometimes contains
+        // children whose position is outside the position of the parent.)
+        all_ranges.sort_by_key(|range| (range.0, std::cmp::Reverse(range.1)));
+        let mut all_ranges = all_ranges.into_iter().peekable();
+
+        // Now we iterate over the tree of ranges to make a disjoint set of ranges.
+        // `stack` is the path in the tree leading to the node that we are currently visiting.
+        let mut stack = Vec::new();
+        let mut disjoint = Vec::new();
+        let mut maybe_cur = all_ranges.next();
+        let mut pos = maybe_cur.as_ref().map(|range| range.0).unwrap_or_default();
+
+        // We accumulate ranges using this closure, which guarantees that they are non-overlapping.
+        // (This is important: RangeMap panics if they are not.)
+        let mut push_range = |end: u32, term: RichTerm| {
+            if pos < end {
+                // range_map::Range has inclusive ends for some reason.
+                // This is overflow-safe because pos < end
+                disjoint.push((Range::new(pos, end - 1), RichTermPtr(term)));
+                pos = end;
+            }
+        };
+
+        while let Some(cur) = maybe_cur {
+            // If the next interval overlaps us, then we must contain it and it is our child.
+            // Otherwise, we are a leaf and it is a sibling or a cousin or something.
+            let next_start = all_ranges.peek().map(|r| r.0);
+            if let (Some(next_start), Some(true)) = (next_start, next_start.map(|x| cur.1 > x)) {
+                // It is our child.
+                push_range(next_start, cur.2.clone());
+                stack.push(cur);
+                maybe_cur = all_ranges.next();
+            } else {
+                // It is a sibling/cousin
+                push_range(cur.1, cur.2.clone());
+                maybe_cur = stack.pop().or_else(|| all_ranges.next());
+            }
+        }
+
+        PositionLookup {
+            table: disjoint.into_iter().collect(),
```

Comment:
Does the table have a notion of "most precise interval", that is we can add e.g. `[0, 10]`, `[0,20]`, `[4,5]` in any order and it ill always give the best match, or does the table rather relies on the order of insertion, such that we insert from the less specific to the most specific one?

If the latter is true, this means that we always have to sort the list of intervals beforehand anyway. In this case, I wonder if it is worth it to pull yet another crate, because I feel we could achieve exactly the same by two call to `binary_search` or something, which would be also logarithmic. Or am I missing something in the picture?

Reply:
The table has disjoint intervals (which we ensure manually, always choosing the most specific one). The crate indeed just uses a binary search, so I'll see about just doing that ourselves.